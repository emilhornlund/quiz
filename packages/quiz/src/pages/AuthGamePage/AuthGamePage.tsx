import React, { FC, useEffect, useRef } from 'react'
import { useNavigate, useSearchParams } from 'react-router-dom'

import { useQuizServiceClient } from '../../api/use-quiz-service-client.tsx'
import { LoadingSpinner, Page } from '../../components'
import { notifyError } from '../../utils/notification.ts'

const AuthGamePage: FC = () => {
  const navigate = useNavigate()

  const [searchParams] = useSearchParams()

  const { authenticateGame } = useQuizServiceClient()

  const hasRequestedRef = useRef(false)

  useEffect(() => {
    const gameId = searchParams.get('id') || undefined
    const gamePIN = searchParams.get('pin') || undefined

    if ((!gameId && !gamePIN) || hasRequestedRef.current) return

    hasRequestedRef.current = true

    authenticateGame({ gameId, gamePIN })
      .then(() => {
        navigate('/join')
      })
      .catch(() => {
        navigate('/')
        notifyError('Game not found. Please try again.')
      })
  }, [searchParams, authenticateGame, navigate])

  return (
    <Page>
      <LoadingSpinner />
    </Page>
  )
}

export default AuthGamePage
