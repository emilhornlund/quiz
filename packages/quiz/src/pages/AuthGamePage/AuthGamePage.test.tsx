import { render, waitFor } from '@testing-library/react'
import React from 'react'
import { MemoryRouter, Route, Routes } from 'react-router-dom'
import { beforeEach, describe, expect, it, vi } from 'vitest'

const h = vi.hoisted(() => ({
  authenticateGame: vi.fn<
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    [{ gameId?: string; gamePIN?: string }],
    Promise<void>
  >(),
  navigate: vi.fn(),
  notifyError: vi.fn(),
}))

vi.mock('../../api/use-quiz-service-client.tsx', () => ({
  useQuizServiceClient: () => ({ authenticateGame: h.authenticateGame }),
}))

vi.mock('react-router-dom', async (orig) => {
  const actual = await orig()
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return { ...actual, useNavigate: () => h.navigate }
})

vi.mock('../../utils/notification.ts', () => ({
  notifyError: h.notifyError,
}))

import AuthGamePage from './AuthGamePage'

describe('AuthGamePage', () => {
  beforeEach(() => {
    h.authenticateGame.mockReset()
    h.navigate.mockReset()
    h.notifyError.mockReset()
  })

  it('does nothing when no id or pin', () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/auth/game']}>
        <Routes>
          <Route path="/auth/game" element={<AuthGamePage />} />
        </Routes>
      </MemoryRouter>,
    )
    expect(h.authenticateGame).not.toHaveBeenCalled()
    expect(h.navigate).not.toHaveBeenCalled()
    expect(container).toMatchSnapshot()
  })

  it('authenticates with gameId and navigates to /join on success', async () => {
    h.authenticateGame.mockResolvedValueOnce({})
    const { container } = render(
      <MemoryRouter initialEntries={['/auth/game?id=abc-123']}>
        <Routes>
          <Route path="/auth/game" element={<AuthGamePage />} />
        </Routes>
      </MemoryRouter>,
    )
    await waitFor(() => {
      expect(h.authenticateGame).toHaveBeenCalledWith({
        gameId: 'abc-123',
        gamePIN: undefined,
      })
    })
    await waitFor(() => expect(h.navigate).toHaveBeenCalledWith('/join'))
    expect(container).toMatchSnapshot()
  })

  it('authenticates with gamePIN and navigates to /join on success', async () => {
    h.authenticateGame.mockResolvedValueOnce({})
    const { container } = render(
      <MemoryRouter initialEntries={['/auth/game?pin=777777']}>
        <Routes>
          <Route path="/auth/game" element={<AuthGamePage />} />
        </Routes>
      </MemoryRouter>,
    )
    await waitFor(() =>
      expect(h.authenticateGame).toHaveBeenCalledWith({
        gameId: undefined,
        gamePIN: '777777',
      }),
    )
    await waitFor(() => expect(h.navigate).toHaveBeenCalledWith('/join'))
    expect(container).toMatchSnapshot()
  })

  it('authenticates with both id and pin and navigates to /join on success', async () => {
    h.authenticateGame.mockResolvedValueOnce({})
    const { container } = render(
      <MemoryRouter initialEntries={['/auth/game?id=abc&pin=123']}>
        <Routes>
          <Route path="/auth/game" element={<AuthGamePage />} />
        </Routes>
      </MemoryRouter>,
    )
    await waitFor(() =>
      expect(h.authenticateGame).toHaveBeenCalledWith({
        gameId: 'abc',
        gamePIN: '123',
      }),
    )
    await waitFor(() => expect(h.navigate).toHaveBeenCalledWith('/join'))
    expect(container).toMatchSnapshot()
  })

  it('navigates home and notifies on failure', async () => {
    h.authenticateGame.mockRejectedValueOnce(new Error('nope'))
    const { container } = render(
      <MemoryRouter initialEntries={['/auth/game?id=missing']}>
        <Routes>
          <Route path="/auth/game" element={<AuthGamePage />} />
        </Routes>
      </MemoryRouter>,
    )
    await waitFor(() => expect(h.navigate).toHaveBeenCalledWith('/'))
    expect(h.notifyError).toHaveBeenCalledWith(
      'Game not found. Please try again.',
    )
    expect(container).toMatchSnapshot()
  })

  it('runs only once even if user clicks back/forward causing re-render with same params', async () => {
    h.authenticateGame.mockResolvedValueOnce({})
    const { container } = render(
      <MemoryRouter initialEntries={['/auth/game?id=one']} initialIndex={0}>
        <Routes>
          <Route path="/auth/game" element={<AuthGamePage />} />
        </Routes>
      </MemoryRouter>,
    )
    await waitFor(() => expect(h.authenticateGame).toHaveBeenCalledTimes(1))
    expect(container).toMatchSnapshot()
  })
})
