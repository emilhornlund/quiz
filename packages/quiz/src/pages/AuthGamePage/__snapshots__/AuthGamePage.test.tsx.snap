// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AuthGamePage > authenticates with both id and pin and navigates to /join on success 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      >
        <a
          data-discover="true"
          href="/auth/login"
        >
          Login
        </a>
      </div>
    </div>
    <div
      class="content centerAlign"
    >
      <div
        class="contentWrapper fullWidth"
      >
        <div
          class="contentInner"
        >
          <div
            class="loadingSpinnerContainer"
            data-testid="loading-spinner"
          >
            <div />
            <div />
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AuthGamePage > authenticates with gameId and navigates to /join on success 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      >
        <a
          data-discover="true"
          href="/auth/login"
        >
          Login
        </a>
      </div>
    </div>
    <div
      class="content centerAlign"
    >
      <div
        class="contentWrapper fullWidth"
      >
        <div
          class="contentInner"
        >
          <div
            class="loadingSpinnerContainer"
            data-testid="loading-spinner"
          >
            <div />
            <div />
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AuthGamePage > authenticates with gamePIN and navigates to /join on success 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      >
        <a
          data-discover="true"
          href="/auth/login"
        >
          Login
        </a>
      </div>
    </div>
    <div
      class="content centerAlign"
    >
      <div
        class="contentWrapper fullWidth"
      >
        <div
          class="contentInner"
        >
          <div
            class="loadingSpinnerContainer"
            data-testid="loading-spinner"
          >
            <div />
            <div />
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AuthGamePage > does nothing when no id or pin 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      >
        <a
          data-discover="true"
          href="/auth/login"
        >
          Login
        </a>
      </div>
    </div>
    <div
      class="content centerAlign"
    >
      <div
        class="contentWrapper fullWidth"
      >
        <div
          class="contentInner"
        >
          <div
            class="loadingSpinnerContainer"
            data-testid="loading-spinner"
          >
            <div />
            <div />
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AuthGamePage > navigates home and notifies on failure 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      >
        <a
          data-discover="true"
          href="/auth/login"
        >
          Login
        </a>
      </div>
    </div>
    <div
      class="content centerAlign"
    >
      <div
        class="contentWrapper fullWidth"
      >
        <div
          class="contentInner"
        >
          <div
            class="loadingSpinnerContainer"
            data-testid="loading-spinner"
          >
            <div />
            <div />
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AuthGamePage > runs only once even if user clicks back/forward causing re-render with same params 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      >
        <a
          data-discover="true"
          href="/auth/login"
        >
          Login
        </a>
      </div>
    </div>
    <div
      class="content centerAlign"
    >
      <div
        class="contentWrapper fullWidth"
      >
        <div
          class="contentInner"
        >
          <div
            class="loadingSpinnerContainer"
            data-testid="loading-spinner"
          >
            <div />
            <div />
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
